pkg/controllers/provisioning/scheduling/topologynodefilter.go

```diff
@@ -27,102 +27,60 @@ import (
+	for _, term := range p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		requirements := nodeSelectorRequirements.Add(v1alpha5.NewRequirements(term.MatchExpressions...).Requirements...)
+		filter = append(filter, requirements)
 	}
 
-	for _, term := range t.nodeSelectorTerms {
-		requirement := v1alpha5.NewRequirements(term.MatchExpressions...)
-		if t.matchesRequirements(requirement, node) {
-			return true
-		}
-	}
+	return filter
+}
 
-	// at this point we have node selector terms, but didn't match all of the requirements for any individual term
-	return false
+// Matches returns true if the TopologyNodeFilter doesn't prohibit node from the participating in the topology
+func (t TopologyNodeFilter) Matches(node *v1.Node) bool {
+	nodeLabels := v1alpha5.NewLabelRequirements(node.Labels)
+	return t.isCompatible(nodeLabels)
 }
 
 // MatchesRequirements returns true if the TopologyNodeFilter doesn't prohibit a node with the requirements from
 // participating in the topology. This method allows checking the requirements from a scheduling.Node to see if the
 // node we will soon create participates in this topology.
-func (t *TopologyNodeFilter) MatchesRequirements(nodeRequirements v1alpha5.Requirements) bool {
-	if t == nil {
-		return true
-	}
-	for k, v := range t.nodeSelector {
-		if !nodeRequirements.Get(k).Has(v) {
-			return false
-		}
-	}
-	if len(t.nodeSelectorTerms) == 0 {
-		return true
-	}
-
-	for _, term := range t.nodeSelectorTerms {
-		requirement := v1alpha5.NewRequirements(term.MatchExpressions...)
-		matchesAllReqs := true
-		for key := range requirement.Keys() {
-			if requirement.Get(key).Intersection(nodeRequirements.Get(key)).Len() == 0 {
-				matchesAllReqs = false
-				break
-			}
-		}
-		// these terms are OR'd together, so if we match one full set of requirements, the filter passes
-		if matchesAllReqs {
-			return true
-		}
-	}
-	return true
+func (t TopologyNodeFilter) MatchesRequirements(nodeRequirements v1alpha5.Requirements) bool {
+	return t.isCompatible(nodeRequirements)
 }
 
-func (t *TopologyNodeFilter) Hash() uint64 {
-	if t == nil || (len(t.nodeSelector) == 0 && len(t.nodeSelectorTerms) == 0) {
+func (t TopologyNodeFilter) Hash() uint64 {
+	if len(t) == 0 {
 		return 0
 	}
-	hash, err := hashstructure.Hash(struct {
-		NodeSelector      map[string]string
-		NodeSelectorTerms []v1.NodeSelectorTerm
-	}{
-		NodeSelector:      t.nodeSelector,
-		NodeSelectorTerms: t.nodeSelectorTerms,
-	}, hashstructure.FormatV2,
+	hash, err := hashstructure.Hash(t, hashstructure.FormatV2,
 		&hashstructure.HashOptions{SlicesAsSets: true})
 	runtime.Must(err)
 	return hash
 
 }
 
-func (t *TopologyNodeFilter) matchesRequirements(requirement v1alpha5.Requirements, node *v1.Node) bool {
-	for key := range requirement.Keys() {
-		if !requirement.Get(key).Has(node.Labels[key]) {
-			return false
+func (t TopologyNodeFilter) isCompatible(requirements v1alpha5.Requirements) bool {
```

Comment:
Thoughts on parity with the Requirements interface?
```suggestion
func (t TopologyNodeFilter) Compatible(requirements v1alpha5.Requirements) bool {
```